Added REST storage for images/signatures resource

miminar opened the pull request:
The new `images/signatures` resource allows to update image' signatures. There's a new images API client call `images.UpdateSignatures()` for this purpose.

A cluster-wide role `system:image-signer` is necessary to add/remove/modify signatures.

A follow-up for #8371

miminar commented:
/cc @mtrmac @deads2k

[test]

soltysh commented:
Will review once the first part is in.

mtrmac commented:
Thanks, I have successfully used this from an independently written client in https://github.com/projectatomic/skopeo/pull/93 ; but I ran into a fairly serious difficulty about the permissions design.

The way things work now, the first user uploading a specific image can include any signatures they want in the original `imagestreammapping` object; any other user uploading the image, even in a different namespace and completely unaware of the first user, has signatures in their `imagestreammapping` ignored, and when they are not a member of the `system:image-signer` role, they can not add their signatures at all.

This is bad for multitenancy or for the stand-alone Atomic Registry; everyone should be able to upload any images they please in their own namespace, and add any signatures they please to images in their own namespace.

It is fine if _removing_ signatures is restricted; adding a signature should never break anything or cause the image to be removed, so removing signatures is not needed except for an “approval/revoking approval” workflow.

Could the permission design be modified so that adding signatures (through the new `/images/`…`/signatures` endpoint at least, possibly also through an `imagestreammapping`) did not require any additional permissions?

(Even with that, the signatures would not be namespaced, and the existence a tenant’s signature would be visible to other tenants. That is a slight privacy leak, but perhaps not a showstopper.)

liggitt commented:
> the first user uploading a specific image can include any signatures they want in the original imagestreammapping object

@miminar, is that intentional?

> Could the permission design be modified so that adding signatures (through the new /images/…/signatures endpoint at least, possibly also through an imagestreammapping)...

We wouldn't let normal users interact with unnamespaced objects in that way. If we wanted regular users to add signatures, there would need to be a namespaced API to allow that.

>  ...did not require any additional permissions?

I would expect it to be a subresource, which would allow granular control over whether signing was enabled. Access to that subresource could possibly be added to existing roles if that makes sense.

deads2k commented:
I think we can resolve many use-cases by having signatures live as a subresource on images.  The status-y parts would be stripped when POSTing to the subresource.  The status-y parts could be set via a direct image update/patch or via a status endpoint (if we split the signature, but right now nothing in an image is split along spec/status lines).  This would mean that image signatures start out as only available to people more trusted that project-admins and editors.

I agree that we'll end up wanting to allow any editor to add a signature to an image, but the "scope" of that signature is less clear.  Because there are size limits for an etcd record, allowing any editor to add any number of signatures won't scale out for large environments.  There are definitely cases where you want to share the signature (QA signs an image and the production namespace should get that sig when they retag) and there are definitely cases where you don't want to share it ("bob signed his istag of the openshift/rhel 7.2, but I don't know who bob is and don't care").  If we stored these signatures in a different etcd hierarchy, we could automatically inject signatures from your namespace.  If you want to see other signatures, maybe you need a different API call with a special filter.

Either way, I think I see the `images/signatures` endpoint as a good starting point for "special" components to use for signing.

miminar commented:
> > the first user uploading a specific image can include any signatures they want in the original imagestreammapping object
>
> @miminar, is that intentional?

It is. An uploader of a new image should be allowed to upload any signatures he wants. New image is tagged just in one namespace so the uploader may be considered as the sole owner of the image at that time.

Modification of signatures of existing image is another story because the image may be tagged in multiple namespaces with different owners. That's why changing the existing signatures is restricted.

> Could the permission design be modified so that adding signatures (through the new /images/…/signatures endpoint at least, possibly also through an imagestreammapping) did not require any additional permissions?

This could be done. I'm however reluctant to allow removal of existing signatures while creating images using `imagestreammapping`. In this case we could just append all the new signatures to the old ones.

However, I don't consider `imagestreammapping` to be the right resource for updating image signatures. It's designed for registry to create a new image entry and tag it into existing image stream at the same time.

For this purpose, I'd rather add a new subresource `imagestreamimages/signatures` or perhaps just add an `AddSignatures()` call to`imagestreamimages`. It would permit just additions. It would be scoped to user's namespace, so he would be able to add signatures to any image tagged into his namespace.

Any signature removal of existing image should be done through the unnamespaced `images/signatures` resource requiring cluster-wide `image-signer` role.

> Either way, I think I see the images/signatures endpoint as a good starting point for "special" components to use for signing.

:+1:

liggitt commented:
> An uploader of a new image should be allowed to upload any signatures he wants. New image is tagged just in one namespace so the uploader may be considered as the sole owner of the image at that time.

I'm not sure whether I'd expect the first person to push an image to be able to surface their signatures into any other namespace they want. @smarterclayton, was that what you expected?

miminar commented:
> I'm not sure whether I'd expect the first person to push an image to be able to surface their signatures into any other namespace they want. @smarterclayton, was that what you expected?

@liggitt shall we rather allow for per-project image signatures? And allow to *pin* a signature to an image by a cluster *image-signer*. I don't disagree. My only concern is signature redundancy and possible complexity.

smarterclayton commented:
With pull through present, it's not sufficient to say "first pusher gets to
push signatures".  For now, only admins can push signatures.  David and I
went through some options here, but for the short term signatures are for
admins and infrastructure, and these APIs are targeted towards that.

On Wed, Jun 8, 2016 at 9:41 AM, Michal Minar <notifications@github.com>
wrote:

> I'm not sure whether I'd expect the first person to push an image to be
> able to surface their signatures into any other namespace they want.
> @smarterclayton <https://github.com/smarterclayton>, was that what you
> expected?
>
> @liggitt <https://github.com/liggitt> shall we rather allow for
> per-project image signatures? And allow to *pin* a signature to an image
> by a cluster *image-signer*. I don't disagree. My only concern is
> signature redundancy and possible complexity.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9181#issuecomment-224591951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p5JpexFFrnZ9rUVBnCdYBx36Mou2ks5qJsZtgaJpZM4IvEFx>
> .
>

mtrmac commented:
>> Either way, I think I see the images/signatures endpoint as a good starting point for "special" components to use for signing.

> :+1:

Yeah, let's get this merged and then figure out the per-namespace aspect; it does need to be figured out eventually.

miminar commented:
Rebased. Removed an option to set signatures during `Create(image)`. Signatures can now be updated only using `images/signatures` subresource.

We can add imagestream-scoped modification of iimage signatures (additions only) in a follow-up PR.

smarterclayton commented:
Let's hold off on adding user focused signing until we have a design
proposal for securing it and limiting the impact on the cluster.  There
were enough questions raised that I'm not sure I want to go forward until
we have agreement.  Let's focus on admin and infra signing for now.

On Mon, Jun 13, 2016 at 4:38 AM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test SUCCESS (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/4751/)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9181#issuecomment-225520415>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_px6o_GqN14L_RUQsfgFzOYto_8dJks5qLRchgaJpZM4IvEFx>
> .
>

deads2k commented on pkg/authorization/api/deprecated_fields.go:
```diff
@@ -63,7 +63,7 @@ var (
 		OpenshiftExposedGroupName:   {BuildGroupName, ImageGroupName, DeploymentGroupName, TemplateGroupName, "routes"},
 		OpenshiftAllGroupName: {OpenshiftExposedGroupName, UserGroupName, OAuthGroupName, PolicyOwnerGroupName, SDNGroupName, PermissionGrantingGroupName, OpenshiftStatusGroupName, "projects",
-			"clusterroles", "clusterrolebindings", "clusterpolicies", "clusterpolicybindings", "images" /* cluster scoped*/, "projectrequests", "builds/details", "imagestreams/secrets",
+			"clusterroles", "clusterrolebindings", "clusterpolicies", "clusterpolicybindings", "images", "images/signatures" /* cluster scoped*/, "projectrequests", "builds/details", "imagestreams/secrets",
```
Don't add anything else to this file.  See line 7.

deads2k commented on pkg/client/images.go:
```diff
@@ -61,3 +62,10 @@ func (c *images) Delete(name string) (err error) {
 }
+
+// UpdateSignatures allows to modify signatures of given image.
+func (c *images) UpdateSignatures(image *imageapi.Image) (result *imageapi.Image, err error) {
```
Seems like this is more `CreateSignatures`.  It's updating an underlying object, but you're actually creating a thing, right?

deads2k replied:
> Seems like this is more CreateSignatures. It's updating an underlying object, but you're actually creating a thing, right?

@smarterclayton I think I'd do a wrapper on `ImageSignature` to register it as an API type that can include the image it wants to use.  Similar to `api.Binding` used for `pods/binding`.  Are you attached to stripped down `Image` instead?  I think its less obvious.

smarterclayton replied:
Yeah I think the wrapper that makes sense and models what we'd do for end
users.  Ref to image, although we may want to make the signature name be
the unique verification (I don't know that I want a special name for a
signature).

On Mon, Jun 13, 2016 at 1:25 PM, David Eads <notifications@github.com>
wrote:

> In pkg/client/images.go
> <https://github.com/openshift/origin/pull/9181#discussion_r66832256>:
>
> > @@ -61,3 +62,10 @@ func (c *images) Delete(name string) (err error) {
> >  	err = c.r.Delete().Resource("images").Name(name).Do().Error()
> >  	return
> >  }
> > +
> > +// UpdateSignatures allows to modify signatures of given image.
> > +func (c *images) UpdateSignatures(image *imageapi.Image) (result *imageapi.Image, err error) {
>
> Seems like this is more CreateSignatures. It's updating an underlying
> object, but you're actually creating a thing, right?
>
> @smarterclayton <https://github.com/smarterclayton> I think I'd do a
> wrapper on ImageSignature to register it as an API type that can include
> the image it wants to use. Similar to api.Binding used for pods/binding.
> Are you attached to stripped down Image instead? I think its less obvious.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9181/files/9ff4bf43548c758b6767b639b335681285fece48#r66832256>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p2jy59h6ZZUdS7qocfm2QYiYn8vHks5qLZKSgaJpZM4IvEFx>
> .
>

deads2k replied:
> Yeah I think the wrapper that makes sense and models what we'd do for end
> users.  Ref to image, although we may want to make the signature name be
> the unique verification (I don't know that I want a special name for a
> signature).

That makes sense.  Does the `RESTHandler` include the requested name part of the URL anywhere in the context to avoid having to repeat myself in the API?

smarterclayton replied:
NamedCreater.  I don't know if someone removed the NamedUpdater equivalent
or not.

On Mon, Jun 13, 2016 at 1:32 PM, David Eads <notifications@github.com>
wrote:

> In pkg/client/images.go
> <https://github.com/openshift/origin/pull/9181#discussion_r66833393>:
>
> > @@ -61,3 +62,10 @@ func (c *images) Delete(name string) (err error) {
> >  	err = c.r.Delete().Resource("images").Name(name).Do().Error()
> >  	return
> >  }
> > +
> > +// UpdateSignatures allows to modify signatures of given image.
> > +func (c *images) UpdateSignatures(image *imageapi.Image) (result *imageapi.Image, err error) {
>
> Yeah I think the wrapper that makes sense and models what we'd do for end
> users. Ref to image, although we may want to make the signature name be
> the unique verification (I don't know that I want a special name for a
> signature).
>
> That makes sense. Does the RESTHandler include the requested name part of
> the URL anywhere in the context to avoid having to repeat myself in the API?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9181/files/9ff4bf43548c758b6767b639b335681285fece48#r66833393>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p1XYyE9Yqlv79b12At_iZl5lg_8Hks5qLZQ_gaJpZM4IvEFx>
> .
>

deads2k replied:
> NamedCreater.  I don't know if someone removed the NamedUpdater equivalent
> or not.

@miminar that's the one you want.  And since its a different type, POST works out very cleanly.

As you expand the storage, I'd like to see a stub switch/case calling out known options for interpreting the signature.

miminar replied:
>> NamedCreater. I don't know if someone removed the NamedUpdater equivalent or not.
>
> @miminar that's the one you want. And since its a different type, POST works out very cleanly.

It makes certainly sense from the API's point of view. However, I'm struggling to find an appropriate format for the signature name:

1. Shall it be a composition of `Type` and `Content`?
  - The `Content` may be quite long - I certainly wouldn't like to be named like that. And the redundancy.
2. Shall we make some digest of `<type>:<content>` instead?
  - That's client unfriendly.
3. So shall it be some auto-generated string?

If we solve that, the `Create` part looks easy. But I'm not sure about signature removal. I can't find anything like `NamedDeleter` (only in upstream comment). Does it mean we have to create it upstream?

```go
func (c *images) CreateSignature(imageName string, signature *imageapi.ImageSignature) (result *imageapi.ImageSignature, err error) {}
func (c *images) DeleteSignature(imageName, name string) (err error) {}
```

deads2k replied:
> I can't find anything like NamedDeleter (only in upstream comment). Does it mean we have to create it upstream?

hehe.  @smarterclayton named subresources? :)

Lacking a named subresource (I doubt @smarterclayton will allow that, but it would be nice), you'll be promoting this to a top level entity if we think that deletion is essential.

You need the name to be 1-to-1 mappable to the signature, probably in the forward direction.  Again, without named subresource, things get weird.  You'll need to include and parse out image.Name, signature.Type, signature.Content into a string we can parse out of a URL.

We could also consider adding a constrained `Name` field to the signature that is settable by callers.  That probably makes more sense.  Then you'll need `image.Name` and `signature.Name` to make a working top-level resource.

mtrmac replied:
(I have only a vague idea about what named subresources and this discussion is about…)

Deleting signatures is not quite essential, but could be very useful.  (We do need at least one of deleting signatures and untagging images, to handle the “image is known to have a vulnerability and is no longer allowed” case.)

If we need a unique identifier, moving this responsibility to clients (who don’t have any way to coordinate setting non-colliding names that is not available to the server; if anything, they have fewer ways to do so) seems to put an undue burden on them.

Wouldn’t it be easier for the API server to assign a UUID on the first upload of a signature? Clients would upload signatures without specifying an UUID, clients who upload a duplicate signature would be silently ignored. The UUIDs would be available when listing signatures (a single API call to get all signatures and their UUIDs, the common case is to need to know all of the signatures), and _only_ used for deleting signatures: Copying an image from one server to another would not preserve these UUIDs.

miminar replied:
> We could also consider adding a constrained Name field to the signature that is settable by callers. 

If I get this correctly the *constraint* will be a name of containing image as a name's prefix. Followed by some separator (e.g. `#`). And the rest will be up to client's decision:

```go
signature.Name = "sha256:be29438d43e16dc0b63fbf64d7f989e85ccbab9546ac4b23ecd385194d0ff675#rhel7-base"
```

I must say I'd still prefer the `UpdateSignatures()` call over adding some randomly generated (though constrained) name to signature. It seems more like an obstacle to API use.

smarterclayton replied:
A signature is a single entity - trying to cram it into UpdateSignatures is
defeating the purpose of it being a unique item.

Named subresources are not allowed currently, and I don't think this
feature crosses the bar to introduce them.  So it has to be a top level
resource.

On Tue, Jun 14, 2016 at 8:59 AM, Michal Minar <notifications@github.com>
wrote:

> In pkg/client/images.go
> <https://github.com/openshift/origin/pull/9181#discussion_r66964209>:
>
> > @@ -61,3 +62,10 @@ func (c *images) Delete(name string) (err error) {
> >  	err = c.r.Delete().Resource("images").Name(name).Do().Error()
> >  	return
> >  }
> > +
> > +// UpdateSignatures allows to modify signatures of given image.
> > +func (c *images) UpdateSignatures(image *imageapi.Image) (result *imageapi.Image, err error) {
>
> We could also consider adding a constrained Name field to the signature
> that is settable by callers.
>
> If I get this correctly the *constraint* will be a name of containing
> image as a name's prefix. Followed by some separator (e.g. #). And the
> rest will be up to client's decision:
>
> signature.Name = "sha256:be29438d43e16dc0b63fbf64d7f989e85ccbab9546ac4b23ecd385194d0ff675#rhel7-base"
>
> I must say I'd still prefer the UpdateSignatures() call over adding some
> randomly generated (though constrained) name to signature. It seems more
> like an obstacle to API use.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9181/files/9ff4bf43548c758b6767b639b335681285fece48#r66964209>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p5HwMlS_HVh1hE3G-IPkv3jaowD-ks5qLqWUgaJpZM4IvEFx>
> .
>

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -393,6 +394,15 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
+				Name: ImageSignerRoleName,
+			},
+			Rules: []authorizationapi.PolicyRule{
+				authorizationapi.NewRule("get").Groups(imageGroup).Resources("images").RuleOrDie(),
```
I'm ok with making an image signer role sufficiently powerful to make it one stop shopping for someone trying to do it, but in that case you should add the ability to pull any image it knows about.

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -570,6 +580,7 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 				authorizationapi.NewRule(readWrite...).Groups(imageGroup).Resources("imagestreamimages", "imagestreammappings", "imagestreams", "imagestreams/secrets", "imagestreamtags").RuleOrDie(),
 				authorizationapi.NewRule("create").Groups(imageGroup).Resources("imagestreamimports").RuleOrDie(),
 				authorizationapi.NewRule("get", "update").Groups(imageGroup).Resources("imagestreams/layers").RuleOrDie(),
+				authorizationapi.NewRule("update").Groups(imageGroup).Resources("images/signatures").RuleOrDie(),
```
images are cluster scoped, which means that images/signatures is also cluster scoped, which means that a RegistryAdmin (project scoped) won't be able to do this.

deads2k commented on pkg/image/api/helper.go:
```diff
@@ -903,3 +904,41 @@ func SetContainerImageEntrypointAnnotation(annotations map[string]string, contai
+}
+
+// RemoveImageSignature removes signature identified by type and blob from the image if it has it.
+func RemoveImageSignature(image *Image, sType string, blob []byte) {
```
Make it an `IndexOfImageSignature` function, then use that value to do a `copy` operation to collapse the list in one go.

deads2k commented on pkg/image/registry/image/etcd/etcd.go:
```diff
@@ -59,8 +59,27 @@ func NewREST(optsGetter restoptions.Getter) (*REST, error) {
+}
+
+// Update alters the signatures subset of an image object.
+func (r *SignatureREST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
```
Obviously, I'd be looking for updates here to match the changes I suggested in the client code.

deads2k commented on pkg/image/registry/image/strategy.go:
```diff
@@ -112,6 +112,37 @@ func (imageStrategy) clearSignatureDetails(image *api.Image) {
+	image := obj.(*api.Image)
+
+	image.Labels = oldImage.Labels
+	// Allow to modify Annotations
```
No, signatures should only affect signatures.  That means building this the other way around.

I think this code gets less error prone if we change the `images/signatures` to take *only* a signature.

deads2k commented on pkg/image/registry/image/strategy.go:
```diff
@@ -112,6 +112,37 @@ func (imageStrategy) clearSignatureDetails(image *api.Image) {
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+// It allows to update only signatures field of the image.
+func (s signatureStrategy) PrepareForUpdate(obj, old runtime.Object) {
```
I'd like to see a validation function that enforces that the only thing this can change is the new signature.

miminar commented:
I've reworked the REST API. Integration test is still wrong, I'll update it later. Other than that this is ready for review.

/cc @deads2k, @soltysh

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -236,6 +236,7 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 				// push and pull images
 				authorizationapi.NewRule("get", "update").Groups(imageGroup).Resources("imagestreams/layers").RuleOrDie(),
 				authorizationapi.NewRule("create").Groups(imageGroup).Resources("imagestreamimports").RuleOrDie(),
+				authorizationapi.NewRule("create", "delete").Groups(imageGroup).Resources("imagesignatures").RuleOrDie(),
```
this is a cluster level resource, so you shouldn't add it here.  It won't have an effect.

miminar replied:
Thanks for reminding. Removed.

deads2k commented on pkg/image/api/helper.go:
```diff
@@ -991,3 +992,57 @@ func SetContainerImageEntrypointAnnotation(annotations map[string]string, contai
+
+// SplitImageSignatureName splits given signature name into image name and signature name.
+func SplitImageSignatureName(imageSignatureName string) (imageName, signatureName string, err error) {
+	segments := strings.SplitN(imageSignatureName, "#", 3)
```
regular split and see what you get

smarterclayton replied:
We use SplitN because regular split can create an unbounded number of
objects.

On Jul 1, 2016, at 8:32 AM, David Eads <notifications@github.com> wrote:

In pkg/image/api/helper.go
<https://github.com/openshift/origin/pull/9181#discussion_r69317140>:

> +func JoinImageSignatureName(imageName, signatureName string) (string, error) {
> +	if len(imageName) == 0 {
> +		return "", fmt.Errorf("imageName may not be empty")
> +	}
> +	if len(signatureName) == 0 {
> +		return "", fmt.Errorf("signatureName may not be empty")
> +	}
> +	if strings.Count(imageName, "#") > 0 || strings.Count(signatureName, "#") > 0 {
> +		return "", fmt.Errorf("neither imageName nor signatureName can contain '#'")
> +	}
> +	return fmt.Sprintf("%s#%s", imageName, signatureName), nil
> +}
> +
> +// SplitImageSignatureName splits given signature name into image name and signature name.
> +func SplitImageSignatureName(imageSignatureName string) (imageName, signatureName string, err error) {
> +	segments := strings.SplitN(imageSignatureName, "#", 3)

regular split and see what you get

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9181/files/bbf390c0306766c27a49c808deddd78621151fcc#r69317140>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p4yLqIeI27FV0wxRi3nDWQTb4Gywks5qRTLkgaJpZM4IvEFx>
.

miminar replied:
Using regular split now.

miminar replied:
> We use SplitN because regular split can create an unbounded number of objects.

That was my reasoning as well, but I didn't want to argue.

deads2k commented on pkg/image/registry/image/etcd/etcd.go:
```diff
@@ -59,8 +59,27 @@ func NewREST(optsGetter restoptions.Getter) (*REST, error) {
+	return &REST{store}, &InternalREST{store: &internalStore}, nil
+}
+
+// InternalREST implements a RESTStorage for image resources internally modifiable.
```
This is a bad name.  What is this thing doing?

miminar replied:
It allows to modify image from internal controllers. Since image is mostly immutable, currently it allows to modify only image signatures. It is now used in image signature rest and will be utilized also once we are able to parse signatures and fill/update their metadata.

deads2k replied:
Why can't I do all that using the current `REST` being returned?  Seems like it has an update function, right?

kargakis replied:
> Why can't I do all that using the current REST being returned? Seems like it has an update function, right?

We should ignore updates on those fields in the normal path.

deads2k replied:
I disagree.  If someone can directly mutate the container (images), why
shouldn't they be allowed to control the signatures?

On Thu, Jul 14, 2016 at 1:55 PM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/image/registry/image/etcd/etcd.go
> <https://github.com/openshift/origin/pull/9181#discussion_r70852033>:
>
> >  	}
> >
> > -	return &REST{store}, nil
> > +	internalStore := *store
> > +	internalStore.UpdateStrategy = image.InternalStrategy
> > +	internalStore.CreateStrategy = nil
> > +
> > +	return &REST{store}, &InternalREST{store: &internalStore}, nil
> > +}
> > +
> > +// InternalREST implements a RESTStorage for image resources internally modifiable.
>
> Why can't I do all that using the current REST being returned? Seems like
> it has an update function, right?
>
> We should ignore updates on those fields in the normal path.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9181/files/b1c344fd4d044ec643a5259b37924dcaff2e7ec3#r70852033>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH2BSjZRauT9y30q26yYBXtGWBdj0MT4ks5qVngegaJpZM4IvEFx>
> .
>

miminar replied:
> I disagree.  If someone can directly mutate the container (images), why shouldn't they be allowed to control the signatures?

As discussed before, we don't want to allow to set signatures during image
push nor image import.

The image rest storage is being used in other virtual rest stores
(ImageStreamMapping, ImageStreamImport). It seems reasonable to me to allow
admin to create/update image directly with any signatures he wants using the
regular store. The other virtual rests should be passed some limited image
rest though.

So instead of adding InternalREST, would it be okay to:

1. add something like UnsignedREST
2. make the image REST fully capable

?

deads2k replied:
> As discussed before, we don't want to allow to set signatures during image
> push nor image import.

We curate those endpoints.  I'm fine with the curated endpoints stripping fields to limit their scope.

Since we're going to loopback to the API server to go through proper admission/defaulting flows (https://github.com/openshift/origin/pull/9181#discussion_r70851489), the images REST endpoint will have the power to make these changes.  The interface there can't perform that restriction, but the REST endpoint calling can do so carefully.  I think is implementation already restricts it sufficiently.

deads2k commented on pkg/image/registry/imagesignature/registry.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+// Registry is an interface for things that know how to store ImageStreamImage objects.
+type Registry interface {
```
We've been skipping these lately.  All access goes through a client from `pkg/client` which hits `RESTStorage` serverside.

miminar replied:
Removed.

deads2k commented on pkg/image/registry/imagesignature/strategy.go:
```diff
@@ -0,0 +1,74 @@
+func (s *strategy) PrepareForCreate(obj runtime.Object) {
+	signature := obj.(*imageapi.ImageSignature)
+
+	signature.Conditions = nil
```
You're going to need a fuzz tests against this method to make sure that all fields except the whitelisted ones are removed.

miminar replied:
Added.

deads2k replied:
> Added.

thanks

deads2k commented on pkg/image/registry/imagesignature/strategy.go:
```diff
@@ -0,0 +1,74 @@
+	signature.IssuedTo = nil
+
+	// generate signature name if it contains just image name
+	imageName, sigName, _ := imageapi.SplitImageSignatureName(signature.Name)
```
I'd like to see this in a validation method.

miminar replied:
Do you refer to https://github.com/openshift/origin/pull/9181/files#diff-fba43349a7c241693d9910314a2864abR84 ?

deads2k commented on pkg/image/registry/imagesignature/strategy.go:
```diff
@@ -0,0 +1,74 @@
+
+	// generate signature name if it contains just image name
+	imageName, sigName, _ := imageapi.SplitImageSignatureName(signature.Name)
+	glog.Infof("imageName=%q, sigName=%q, signature.Name=%q", imageName, sigName, signature.Name)
```
This glog shouldn't be here.

miminar replied:
Removed.

deads2k commented on pkg/image/registry/imagesignature/strategy.go:
```diff
@@ -0,0 +1,74 @@
+	if len(imageName) == 0 || len(sigName) > 0 {
+		return
+	}
+	sigName = fmt.Sprintf("%x", sha1.Sum([]byte(fmt.Sprintf("%s:%s", signature.Type, signature.Content))))
```
I'm not clear on why we're stomping the name with this.  I thought we decided that users had to choose names for these things so they'd later be able to delete them.  This makes the name unpredictable.

deads2k replied:
> This makes the name unpredictable.

Unpredictable from a user perspective, I know its deterministic, but I'll never remember it.

miminar replied:
User needs to set a full name now.

deads2k commented on pkg/image/registry/imagesignature/rest.go:
```diff
@@ -0,0 +1,93 @@
+	if byName, byContent := imageapi.IndexOfImageSignatureByName(image.Signatures, signature.Name), imageapi.IndexOfImageSignature(image.Signatures, signature.Type, signature.Content); byName >= 0 || byContent >= 0 {
+		return nil, kapierrors.NewAlreadyExists(imageapi.Resource("imageSignatures"), signature.Name)
+	}
+
```
This looks like the spot to perform a check of the signature.  We could do it here before doing it inside of a controller.

miminar replied:
> This looks like the spot to perform a check of the signature. We could do it here before doing it inside of a controller.

See the updated code if it matches your expectations.

miminar commented:
Items addressed.

soltysh commented on pkg/client/imagesignatures.go:
```diff
@@ -0,0 +1,41 @@
+package client
```
@deads2k since @mfojtik's client generation is in aren't we switching slowly to use that one instead of introducing new ones? Especially that this is a new type.

soltysh replied:
Even more, there's a desire upstream to deprecate and remove the manual clients.

deads2k replied:
> Even more, there's a desire upstream to deprecate and remove the "manual" clients.

Turns out that client generation doesn't understand restricted resource, so the generation would produce invalid methods for this resource.

soltysh replied:
OK, we'll need to fix it. But not in this PR, obviously.

soltysh commented:
One question regarding client side, but overall PR LGTM.

miminar commented:
@deads2k may this have another careful look of yours? ... please

deads2k commented on pkg/api/serialization_test.go:
```diff
@@ -187,6 +187,11 @@ func fuzzInternalObject(t *testing.T, forVersion unversioned.GroupVersion, item
 			j.DockerImageMetadataVersion = []string{"pre012", "1.0"}[c.Rand.Intn(2)]
 			j.DockerImageReference = c.RandString()
 		},
+		func(j *image.ImageSignature, c fuzz.Continue) {
```
why does this need fields excluded from fuzzing?  You should also do it, `c.FuzzNoCustom` and then clear the fields you can't set.

miminar replied:
> why does this need fields excluded from fuzzing? You should also do it, c.FuzzNoCustom and then clear the fields you can't set.

fixed

deads2k commented on pkg/image/api/helper.go:
```diff
@@ -1001,3 +1002,57 @@ func SetContainerImageEntrypointAnnotation(annotations map[string]string, contai
+	case 2:
+		signatureName = segments[1]
+		fallthrough
+	case 1:
```
No need for this check. If you hit this directly, you were returning an error.  Just make the `default` case say that `imageName#imageSignature` is the required format.

deads2k commented on pkg/image/api/validation/validation.go:
```diff
@@ -113,9 +128,10 @@ func validateImageSignature(signature *api.ImageSignature, fldPath *field.Path)
-	result := validation.ValidateObjectMetaUpdate(&newImage.ObjectMeta, &oldImage.ObjectMeta, field.NewPath("metadata"))
-	result = append(result, ValidateImage(newImage)...)
+// ValidateImageSignatureUpdate ensures that the new ImageSignature is valid.
+func ValidateImageSignatureUpdate(newImageSignature, oldImageSignature *api.ImageSignature) field.ErrorList {
```
Updating an image shouldn't be allowed to change its type or content

deads2k commented on pkg/image/api/validation/validation.go:
```diff
@@ -67,9 +67,24 @@ func validateImage(image *api.Image, fldPath *field.Path) field.ErrorList {
+
 func validateImageSignature(signature *api.ImageSignature, fldPath *field.Path) field.ErrorList {
-	allErrs := field.ErrorList{}
+	allErrs := validation.ValidateObjectMeta(&signature.ObjectMeta, false, oapi.MinimalNameRequirements, fldPath.Child("metadata"))
```
Let's start by disallowing labels and annotations.  I can think of use-cases for them, but I think it will complicate things unnecessarily to start.

miminar replied:
Labels and annotations disallowed.

deads2k commented on pkg/image/registry/image/strategy.go:
```diff
@@ -38,8 +38,8 @@ func (s imageStrategy) PrepareForCreate(obj runtime.Object) {
 	}
-
-	s.clearSignatureDetails(newImage)
+	// allow to set the signatures only through signatures subresource
```
Why would I want to force this restriction?  Signatures aren't secret things, they are replayable by design.

miminar replied:
Repasting @liggitt's earlier comment here:

> I'm not sure whether I'd expect the first person to push an image to be able to surface their signatures into any other namespace they want. @smarterclayton, was that what you expected?

And re-pasting my earlier comment as well:

> As discussed before, we don't want to allow to set signatures during image
push nor image import.
>
> The image rest storage is being used in other virtual rest stores (ImageStreamMapping, ImageStreamImport). It seems reasonable to me to allow admin to create/update image directly with any signatures he wants using the regular store. The other virtual rests should be passed some limited image rest though.
>
> So instead of adding InternalREST, would it be okay to:
>
>  1. add something like UnsignedREST
>  2. make the image REST fully capable

deads2k replied:
> Repasting @liggitt's earlier comment here:
> 
> I'm not sure whether I'd expect the first person to push an image to be able to surface their signatures into any other namespace they want. @smarterclayton, was that what you expected?

The cluster-admin is the cluster-admin.  If he wants the signatures, he can have them.

If we want to argue about expectations during create and stripping the field during create as a general rule for usability, I could probably see that argument, but that means that updates can still apply them and we can still loopback through the API server to make it work.

deads2k commented on pkg/image/registry/imagesignature/rest.go:
```diff
@@ -0,0 +1,98 @@
+
+// REST implements the RESTStorage interface for ImageSignature
+type REST struct {
+	imageRegistry image.Registry
```
I think I'd have this loopback to the API server to ensure we go back through the full admission/defaulting/validation chain.  @smarterclayton

smarterclayton replied:
Yes, use a client.  It's an infrequent API, it's admin only, and you're only updating the image itself, not status, right?

miminar replied:
> Yes, use a client. It's an infrequent API, it's admin only, and you're only updating the image itself, not status, right?

The rest doesn't touch any image stream. Just the image which has no status.
Using the client sounds good to me. I'll rewrite.

miminar replied:
> I think I'd have this loopback to the API server to ensure we go back through the full admission/defaulting/validation chain.

The downside is that [ImageInterface](https://github.com/miminar/origin/blob/image-signatures-rest/pkg/client/images.go#L15) will have to be extended for `Update()` method.

deads2k replied:
> The downside is that ImageInterface will have to be extended for Update() method.

That's just a client library.  `curl` can already make these call, so it isn't increasing API surface area.

miminar replied:
> That's just a client library. curl can already make these call, so it isn't increasing API surface area.

Still so much to learn...

deads2k commented on pkg/image/registry/imagesignature/strategy_test.go:
```diff
@@ -0,0 +1,91 @@
+
+			j.Content = []byte(c.RandString())
+			for i := 0; i < c.Rand.Intn(3)+2; i++ {
+				j.Labels[c.RandString()] = c.RandString()
```
why clear these just to re-fuzz them?

miminar replied:
> why clear these just to re-fuzz them?

I'm not sure, how the fuzzer work. But nearly every time I run it without setting the labels myself, there were no labels set.

deads2k commented on pkg/image/registry/imagesignature/rest.go:
```diff
@@ -0,0 +1,98 @@
+
+	imageName, _, err := imageapi.SplitImageSignatureName(signature.Name)
+	if err != nil {
+		return nil, kapierrors.NewBadRequest("ImageSignature name must be composed of <imageName>#<signatureName>")
```
`NewBadRequest(err.Error())`?

kargakis replied:
This will be cleaner if handled by validation.

miminar replied:
> This will be cleaner if handled by validation.

It's already handled by validation called from `BeforeCreate` few lines above. So this case shouldn't actually happen.

kargakis commented on pkg/image/api/install/install.go:
```diff
@@ -95,7 +95,7 @@ func addVersionsToScheme(externalVersions ...unversioned.GroupVersion) {
 
 func newRESTMapper(externalVersions []unversioned.GroupVersion) meta.RESTMapper {
-	rootScoped := sets.NewString("Image")
+	rootScoped := sets.NewString("Image", "ImageSignature")
```
Got any ideas for shortcut?

deads2k replied:
> Got any ideas for shortcut?

This is infrequent and most retrieval will usually be done through an istag.  I don't see a need.

deads2k commented on pkg/image/registry/imagesignature/rest.go:
```diff
@@ -0,0 +1,98 @@
+		return nil, kapierrors.NewAlreadyExists(imageapi.Resource("imageSignatures"), signature.Name)
+	}
+
+	if err := validation.ValidateImageSignature(signature); len(err) > 0 {
```
`BeforeCreate` already calls `Validate`

miminar replied:
> Ahh, thanks for remining me.

Ahh, thanks for remining me.

deads2k commented on test/integration/imagesigning_test.go:
```diff
@@ -0,0 +1,130 @@
+	testserver "github.com/openshift/origin/test/util/server"
+)
+
+func TestImageAddAndRemoveSignature(t *testing.T) {
```
I'd like to see coverage of:
 1. delete signature that doesn't exist
 1. delete correct signature of at least two signatures
 1. create signature conflicting by content, but not name

miminar replied:
Added all the test cases requested.

deads2k commented:
several comments, but this is actually pretty close API-wise.  @smarterclayton

kargakis commented on pkg/image/registry/image/strategy.go:
```diff
@@ -129,6 +129,38 @@ func (imageStrategy) clearSignatureDetails(image *api.Image) {
+	image.DockerImageManifestMediaType = oldImage.DockerImageManifestMediaType
+	image.DockerImageConfig = oldImage.DockerImageConfig
+
+	s.clearSignatureDetails(image)
```
Can you add a comment for why you are doing this?

deads2k replied:
> Can you add a comment for why you are doing this?

I think all the `internalREST` stuff goes away once we loopback to the API server instead of using a direct registry connection.

miminar replied:
> Can you add a comment for why you are doing this?

Yes, will do so.

> I think all the internalREST stuff goes away once we loopback to the API server instead of using a direct registry connection.

This should stay (at least in the other strategy) because we're not able to parse signatures yet to set signature metadata fields.

smarterclayton commented on pkg/image/registry/image/strategy.go:
```diff
@@ -129,6 +129,38 @@ func (imageStrategy) clearSignatureDetails(image *api.Image) {
+
+	image.Labels = oldImage.Labels
+
+	image.DockerImageReference = oldImage.DockerImageReference
```
This is unsafe because you could unintentionally forget copying something new.  Do it where only things explicitly known are preserved in new.

miminar replied:
I've removed `internalStrategy`.

deads2k replied:
He wrote a fuzz test to cover them.

On Mon, Jul 18, 2016 at 8:14 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/image/registry/image/strategy.go
> <https://github.com/openshift/origin/pull/9181#discussion_r71254118>:
>
> > +// InternalStrategy is the default logic that applies when updating
> > +// image' signatures via the REST API.
> > +var InternalStrategy = internalStrategy{Strategy}
> > +
> > +// Canonicalize normalizes the object after validation.
> > +func (internalStrategy) Canonicalize(obj runtime.Object) {
> > +}
> > +
> > +// PrepareForUpdate currently allows to set just image signatures.
> > +func (s internalStrategy) PrepareForUpdate(obj, old runtime.Object) {
> > +	oldImage := old.(*api.Image)
> > +	image := obj.(*api.Image)
> > +
> > +	image.Labels = oldImage.Labels
> > +
> > +	image.DockerImageReference = oldImage.DockerImageReference
>
> This is unsafe because you could unintentionally forget copying something
> new. Do it where only things explicitly known are preserved in new.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9181/files/ec22b767710a7fb1cf5496bb96ea668412df8f59#r71254118>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH2BSi2UphCDwKfloPBKFOlWLqK70b5Nks5qXBbzgaJpZM4IvEFx>
> .
>

miminar commented on pkg/cmd/server/origin/master.go:
```diff
@@ -467,6 +468,7 @@ func (c *MasterConfig) GetRestStorage() map[string]rest.Storage {
 	imageStorage, err := imageetcd.NewREST(c.RESTOptionsGetter)
 	checkStorageErr(err)
 	imageRegistry := image.NewRegistry(imageStorage)
+	imageSignatureStorage := imagesignature.NewREST(c.PrivilegedLoopbackOpenShiftClient.Images())
```
@deads2k is this the kind of client you have in mind?

deads2k replied:
> @deads2k is this the kind of client you have in mind?

Yes.

miminar commented on pkg/image/api/v1/types.go:
```diff
@@ -57,41 +57,44 @@ type ImageLayer struct {
+
 	// Required: Describes a type of stored blob.
-	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	Type string `json:"type" protobuf:"bytes,2,opt,name=type"`
```
Can I safely assume that signatures haven't been really used anywhere until now to bump these numbers? Or do I have to increase the index of `ObjectMeta` instead?

deads2k replied:
> Can I safely assume that signatures haven't been really used anywhere until now to bump these numbers? Or do I have to increase the index of ObjectMeta instead?

bumpable

miminar commented:
Flake #9929 and yum flake. re-[test]
```
FAILURE after 60.119s: test/cmd/quota.sh:25: executing 'oc describe appliedclusterresourcequota/for-deads-by-annotation -n bar --as deads' expecting any result and text 'secrets.*18'; re-trying every 0.2s until completion or 60.000s: the command timed out
```

miminar commented:
Flake #9364. re-[test]

miminar commented:
@deads2k is there anything else I can fix?

deads2k commented:
api lgtm.

@smarterclayton any comment?  I'd like to merge it for him tomorrow.

smarterclayton commented:
API is ok except for using # - we need to pick a different character for safety.

miminar commented:
> API is ok except for using # - we need to pick a different character for safety.

Changed to `@`.

deads2k commented:
lgtm [merge]

miminar commented:
@deads2k, @mtrmac thank you! I'm glad we came to settlement ... finally :smile:

Thanks to all the involved for valuable input.
